packages/verkle/src/verkleTree.ts

```diff
@@ -209,193 +412,150 @@ export class VerkleTree {
+    const result: Path = {
+      node: null,
+      stack: [],
+      remaining: key,
+    }
+    if (equalsBytes(this.root(), this.EMPTY_TREE_ROOT)) return result
 
-      // walk tree and process nodes
-      try {
-        await this.walkTree(this.root(), onFound)
-      } catch (error: any) {
-        if (error.message === 'Missing node in DB' && !throwIfMissing) {
-          // pass
-        } else {
-          reject(error)
-        }
-      }
+    // Get root node
+    let rawNode = await this._db.get(ROOT_DB_KEY)
+    if (rawNode === undefined)
+      throw new Error('root node should exist when root not empty tree root')
 
-      // Resolve if walkTree finishes without finding any nodes
-      resolve({ node: null, remaining: new Uint8Array(0), stack })
-    })
-  }
+    const rootNode = decodeNode(rawNode, 0, this.verkleCrypto) as InternalNode
 
-  /**
-   * Walks a tree until finished.
-   * @param root
-   * @param onFound - callback to call when a node is found. This schedules new tasks. If no tasks are available, the Promise resolves.
-   * @returns Resolves when finished walking tree.
-   */
-  async walkTree(root: Uint8Array, onFound: FoundNodeFunction): Promise<void> {
-    await WalkController.newWalk(onFound, this, root)
-  }
+    this.DEBUG &&
+      this.debug(`Starting with Root Node: [${bytesToHex(rootNode.hash())}]`, ['FIND_PATH'])
+    result.stack.push([rootNode, ROOT_DB_KEY])
+    let child = rootNode.children[key[0]]
 
-  /**
-   * Tries to find the leaf node leading up to the given key, or null if not found.
-   * @param key - the search key
-   * @param throwIfMissing - if true, throws if any nodes are missing. Used for verifying proofs. (default: false)
-   */
-  async findLeafNode(key: Uint8Array, throwIfMissing = false): Promise<LeafNode | null> {
-    const { node } = await this.findPath(key, throwIfMissing)
-    if (!(node instanceof LeafNode)) {
-      if (throwIfMissing) {
-        throw new Error('leaf node not found')
+    // Root node doesn't contain a child node's commitment on the first byte of the path so we're done
+    if (equalsBytes(child.commitment, this.verkleCrypto.zeroCommitment)) {
+      this.DEBUG && this.debug(`Partial Path ${key[0]} - found no child.`, ['FIND_PATH'])
+      return result
+    }
+    let finished = false
+    while (!finished) {
+      rawNode = await this._db.get(child.path)
+      // We should always find the node if the path is specified in child.path
+      if (rawNode === undefined) throw new Error(`missing node at ${bytesToHex(child.path)}`)
+      const decodedNode = decodeNode(rawNode, result.stack.length, this.verkleCrypto)
+
+      // Calculate the index of the last matching byte in the key
+      const matchingKeyLength = matchingBytesLength(key, child.path)
+      const foundNode = equalsBytes(key, child.path)
+      if (foundNode || child.path.length >= key.length || decodedNode instanceof LeafNode) {
```

Comment:
Is the last condition necessary? If I'm thinking about this correctly, whenever the child path length is longer than the key, the found node would necessarily be a leaf node?